fix: Make KOptListMoveSelector work when moveThreadCount != NONE

Christopher-Chianelli opened the pull request:
The issue came from the fact MultipleDelegateList returns a new instance of the rebased list on every call. This is problematic, since each FlipSublistAction expects rebase to return the same list across mulitple calls. When different instances are used, the FlipSublistAction modify different lists, causing the Undo move to be incorrect. (Since the list the FlipSublistAction operation on only has its flip incorporated in it instead of all flips). Since a new MultipleDelegateList is created for each move, we can safely store a cached result in one of its fields.

Fixes #250

Christopher-Chianelli committed:
fix: Make KOptListMoveSelector work when moveThreadCount != NONE

The issue came from the fact MultipleDelegateList returns a new
instance of the rebased list on every call. This is problematic,
since each FlipSublistAction expects rebase to return the same
list across mulitple calls. When different instances are used,
the FlipSublistAction modify different lists, causing the Undo
move to be incorrect. (Since the list the FlipSublistAction operation
on only has its flip incorporated in it instead of all flips).
Since a new MultipleDelegateList is created for each move, we can
safely store a cached result in one of its fields.

triceo commented on core/core-impl/src/main/java/ai/timefold/solver/core/impl/heuristic/selector/move/generic/list/kopt/MultipleDelegateList.java:
```diff
@@ -42,6 +43,7 @@ public MultipleDelegateList(List<T>... delegates) {
         }
 
         this.totalSize = sizeSoFar;
+        this.cachedRebaseResult = null;
```
Maybe add a comment here as to why it's safe to store this here?

Christopher-Chianelli replied:
I think it not safe, due to the fact the accepted move is rebased twice (one for it score director, and one for the solver score director); I can move the cached result to KOptListMove

triceo replied:
Go for it. I think it makes more sense to store it there, because that is where `rebase` is.

triceo reviewed (approved):
Good to see the extra coverage.
One minor comment.

Christopher-Chianelli commented:
There seem to be another issue that need to be patched (it fix one problem, but there appears to be another)

Christopher-Chianelli commented:
At this point, I think supplies/internals might be corrupted in multithreaded environment.

First, items in the list returned by SingletonInverseVariableSupply are different:
```
Before: 565809328:[TimeWindowedCustomer-39, TimeWindowedCustomer-72, TimeWindowedCustomer-44, TimeWindowedCustomer-38, TimeWindowedCustomer-86, TimeWindowedCustomer-84, TimeWindowedCustomer-42, TimeWindowedCustomer-92, TimeWindowedCustomer-98, TimeWindowedCustomer-61, *TimeWindowedCustomer-46, TimeWindowedCustomer-97, TimeWindowedCustomer-43, TimeWindowedCustomer-97, TimeWindowedCustomer-91, TimeWindowedCustomer-98, TimeWindowedCustomer-58*]
After: 565809328: [TimeWindowedCustomer-39, TimeWindowedCustomer-72, TimeWindowedCustomer-44, TimeWindowedCustomer-38, TimeWindowedCustomer-86, TimeWindowedCustomer-84, TimeWindowedCustomer-42, TimeWindowedCustomer-92, TimeWindowedCustomer-98, TimeWindowedCustomer-61, *TimeWindowedCustomer-85, TimeWindowedCustomer-97, TimeWindowedCustomer-43, TimeWindowedCustomer-100, TimeWindowedCustomer-91, TimeWindowedCustomer-93, TimeWindowedCustomer-58*]
```

Although it fairly deep: in one case, an item completely disappeared from existing (checked using my only InverseSupply):

```java
        findDelegate: for (int i = 0; i < delegates.length; i++) {
            List<T> delegate = delegates[i];
            if (delegate.isEmpty()) {
                out[i] = new ArrayList<>();
                continue;
            }
            EntityDescriptor<?> entityDescriptor = variableDescriptor.getEntityDescriptor();
            for (Object item : ((SolutionDescriptor<Object>) (entityDescriptor.getSolutionDescriptor()))
                    .getAllEntitiesAndProblemFacts(destinationScoreDirector.getWorkingSolution())) {
                if (entityDescriptor.getEntityClass().isAssignableFrom(item.getClass())) {
                    List<T> rebasedDelegate = (List<T>) variableDescriptor.getListVariable(item);
                    if (rebasedDelegate.contains(destinationScoreDirector.lookUpWorkingObject(delegate.get(0)))) {
                        out[i] = rebasedDelegate;
                        continue findDelegate;
                    }
                }
            }
           // this code was reached, which should never happen, considering it literally iterated through every entity and the solution is initialized.
            throw new IllegalArgumentException(
                    "Cannot find entity corresponding to " + destinationScoreDirector.lookUpWorkingObject(delegate.get(0)));
        }
```

Christopher-Chianelli commented:
Confirmed it was the supplies. I managed to remove the supplies (by realizing I already have the original entities and can lookup list directly), and that fixed the code.

Christopher-Chianelli committed:
fix: Make KOptListMove work in multithreaded environments

The only functional change in this commit is changing rebase from using
`getvar(inverse(lookup(value)))` to using `getvar(lookup(entity))` instead.
Despite the two above expressions being equivalent, only the second
works in multithreaded environments. This leads me to believe
the ListInverseVariableSupply get corrupted in multithreaded
environments.

Christopher-Chianelli commented:
Tested locally, new commit works on FULL_ASSERT multi-threaded vehicle routing with 3-opts

triceo commented:
Thank you @Christopher-Chianelli, good job!
